sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -906,4 +906,45 @@ object DateTimeUtils {
     val days = period.getDays
     new CalendarInterval(months, days, 0)
   }
+
+  /**
+   * Since the Proleptic Gregorian calendar is de-facto calendar worldwide, as well as the chosen
+   * one in ANSI SQL standard, Spark 3.0 switches to it by using DateTimeFormatter classes.
+   * However, the breaking changes between SimpleDateFormat and DateTimeFormatter pattern string
+   * will also breaks the backward compatibility of Spark 2.4 and earlier when parsing datetime.
+   * This function converts all incompatible pattern for the new parser in Spark 3.0. See more
+   * details in SPARK-31030.
+   *
+   * @param pattern The input pattern.
+   * @return The pattern for new parser
+   */
+  def convertIncompatiblePattern(pattern: String): String = {
+    val eraDesignatorContained = pattern.split("'").zipWithIndex.exists {
+      case (patternPart, index) =>
+        // Text can be quoted using single quotes, we only check the non-quote parts.
+        index % 2 == 0 && patternPart.contains("G")
+    }
+    pattern.split("'").zipWithIndex.map {
+      case (patternPart, index) =>
+        if (index % 2 == 0) {
+          // The meaning of 'u' was day number of week in Java 7, it was changed to year in Java 8.
```

Comment:
Java 8 -> DateTimeFormatter

Reply:
Thanks, also rephrase the whole comment.